File: CBLClient/Apps/JavaRequestHandlers/src/main/java/com/couchbase/mobiletestkit/javacommon/RequestHandler/LoggingRequestHandler.java
package com.couchbase.mobiletestkit.javacommon.RequestHandler;

import java.io.File;

import com.couchbase.lite.logging.FileLogSink;
import com.couchbase.lite.logging.LogSinks;
import com.couchbase.mobiletestkit.javacommon.Args;
import com.couchbase.mobiletestkit.javacommon.RequestHandlerDispatcher;
import com.couchbase.mobiletestkit.javacommon.RawData;
import com.couchbase.mobiletestkit.javacommon.util.Log;
import com.couchbase.mobiletestkit.javacommon.util.ZipUtils;

import com.couchbase.lite.*;

public class LoggingRequestHandler {
    private static final String TAG = "LOGREQHANDLER";
    /* ----------- */
    /* - Logging - */
    /* ----------- */

    public FileLogSink configure(Args args) {
        String log_level = args.get("log_level");
        String directory = args.get("directory");
        int maxRotateCount = args.get("max_rotate_count");
        long maxSize = args.get("max_size");
        boolean plainText = args.get("plain_text");

        if (directory.isEmpty()) {
            long ts = System.currentTimeMillis() / 1000;
            directory = RequestHandlerDispatcher.context.getFilesDir().getAbsolutePath() + File.separator + "logs_" + ts;
            Log.i(TAG, "File logging configured at: " + directory);
        }
        FileLogSink.Builder builder = new FileLogSink.Builder()
                .setDirectory(directory);
        if (maxRotateCount > 1) {
            builder.setMaxKeptFiles(maxRotateCount+1);
        }
        if (maxSize > 512000) {
            builder.setMaxFileSize(maxSize);
        }
        builder.setPlainText(plainText);

        LogLevel level = parseLogLevel(log_level);
        builder.setLevel(level);

        FileLogSink sink = builder.build();
        LogSinks.get().setFile(sink);
        return sink;
    }

    public boolean getPlainTextStatus(Args args) {
        FileLogSink sink = LogSinks.get().getFile();
        return sink != null && sink.isPlainText();
    }

    public int getMaxRotateCount(Args args) {
        FileLogSink sink = LogSinks.get().getFile();
        return sink != null ? sink.getMaxKeptFiles() - 1 : 0;
    }

    public long getMaxSize(Args args) {
        FileLogSink sink = LogSinks.get().getFile();
        return sink != null ? sink.getMaxFileSize(): 0;
    }

    public String getDirectory(Args args) {
        FileLogSink sink = LogSinks.get().getFile();
        return sink != null ? sink.getDirectory() : "";
    }

    public FileLogSink getConfig(Args args) {
        return LogSinks.get().getFile();
    }

    public FileLogSink setPlainTextStatus(Args args) {
        Boolean plain_text = args.get("plain_text");
        FileLogSink currentSink = LogSinks.get().getFile();

        if (currentSink == null) return null;

        FileLogSink newSink = new FileLogSink.Builder(currentSink)
                .setPlainText(plain_text)
                .build();

        LogSinks.get().setFile(newSink);
        return newSink;
    }

    public FileLogSink setMaxRotateCount(Args args) {
        int max_rotate_count = args.get("max_rotate_count");
        FileLogSink currentSink = LogSinks.get().getFile();

        if (currentSink == null) return null;

        FileLogSink newSink = new FileLogSink.Builder(currentSink)
                .setMaxKeptFiles(max_rotate_count+1)
                .build();

        LogSinks.get().setFile(newSink);
        return newSink;
    }


    public FileLogSink setMaxSize(Args args) {
        long max_size = args.get("max_size");
        FileLogSink currentSink = LogSinks.get().getFile();

        if (currentSink == null) return null;

        FileLogSink newSink = new FileLogSink.Builder(currentSink)
                .setMaxFileSize(max_size)
                .build();

        LogSinks.get().setFile(newSink);
        return newSink;
    }

    public FileLogSink setConfig(Args args) {
        String directory = args.get("directory");
        if (directory.isEmpty()) {
            long ts = System.currentTimeMillis() / 1000;
            directory = RequestHandlerDispatcher.context.getFilesDir().getAbsolutePath() + File.separator + "logs_" + ts;

            Log.i(TAG, "File logging configured at: " + directory);
        }
        FileLogSink newSink = new FileLogSink.Builder()
                .setDirectory(directory)
                .build();
        LogSinks.get().setFile(newSink);
        return newSink;
    }

    public int getLogLevel(Args args) {
        FileLogSink sink = LogSinks.get().getFile();
        return sink != null ? sink.getLevel().ordinal() : LogLevel.NONE.ordinal();
    }

    public FileLogSink setLogLevel(Args args) {
        String log_level = args.get("log_level");
        FileLogSink currentSink = LogSinks.get().getFile();

        if (currentSink == null) return null;

        LogLevel level = parseLogLevel(log_level);
        FileLogSink newSink = new FileLogSink.Builder(currentSink)
                .setLevel(level)
                .build();

        LogSinks.get().setFile(newSink);
        return newSink;
    }

    public RawData getLogsInZip(Args args) {
        FileLogSink fileLoggerConfig = LogSinks.get().getFile();
        if (fileLoggerConfig == null) { return null; }

        ZipUtils zipper = new ZipUtils();

        File zipDir = RequestHandlerDispatcher.context.getExternalFilesDir("zip");
        try {
            File zipFile = new File(zipDir, "archive.zip");
            if (zipFile.exists()) { zipper.deleteRecursive(zipFile); }

            zipper.zipDirectory(fileLoggerConfig.getDirectory(), zipFile);
            return new RawData("application/zip", zipper.readFile(zipFile));
        }
        finally {
            zipper.deleteRecursive(zipDir);
        }
    }

    private LogLevel parseLogLevel(String log_level) {
        switch (log_level) {
            case "debug": return LogLevel.DEBUG;
            case "verbose": return LogLevel.VERBOSE;
            case "info": return LogLevel.INFO;
            case "error": return LogLevel.ERROR;
            case "warning": return LogLevel.WARNING;
            default: return LogLevel.NONE;
        }
    }

}

File: CBLClient/Apps/JavaRequestHandlersKotlin/src/main/java/com/couchbase/mobiletestkit/javacommon/RequestHandler/LoggingRequestHandler.kt
package com.couchbase.mobiletestkit.javacommon.RequestHandler

import com.couchbase.lite.Database
import com.couchbase.lite.LogFileConfiguration
import com.couchbase.lite.LogLevel
import com.couchbase.mobiletestkit.javacommon.Args
import com.couchbase.mobiletestkit.javacommon.RawData
import com.couchbase.mobiletestkit.javacommon.RequestHandlerDispatcher
import com.couchbase.mobiletestkit.javacommon.util.Log
import com.couchbase.mobiletestkit.javacommon.util.ZipUtils
import java.io.File

class LoggingRequestHandler {
    /* ----------- */ /* - Logging - */ /* ----------- */
    fun configure(args: Args): LogFileConfiguration {
        val log_level = args.get<String>("log_level")
        var directory = args.get<String>("directory")
        val maxRotateCount = args.get<Int>("max_rotate_count")
        val maxSize = args.get<Long>("max_size")
        val plainText = args.get<Boolean>("plain_text")
        if (directory.isEmpty()) {
            val ts = System.currentTimeMillis() / 1000
            directory =
                RequestHandlerDispatcher.context.filesDir.absolutePath + File.separator + "logs_" + ts
            Log.i(
                TAG,
                "File logging configured at: $directory"
            )
        }
        val config = LogFileConfiguration(directory)
        if (maxRotateCount > 1) {
            config.maxRotateCount = maxRotateCount
        }
        if (maxSize > 512000) {
            config.maxSize = maxSize
        }
        config.setUsePlaintext(plainText)
        Database.log.file.config = config
        when (log_level) {
            "debug" -> Database.log.file.level = LogLevel.DEBUG
            "verbose" -> Database.log.file.level = LogLevel.VERBOSE
            "info" -> Database.log.file.level = LogLevel.INFO
            "error" -> Database.log.file.level = LogLevel.ERROR
            "warning" -> Database.log.file.level = LogLevel.WARNING
            else -> Database.log.file.level = LogLevel.NONE
        }
        return config
    }

    fun getPlainTextStatus(args: Args?): Boolean {
        return Database.log.file.config!!.usesPlaintext()
    }

    fun getMaxRotateCount(args: Args?): Int {
        return Database.log.file.config!!.maxRotateCount
    }

    fun getMaxSize(args: Args?): Long {
        return Database.log.file.config!!.maxSize
    }

    fun getDirectory(args: Args?): String {
        return Database.log.file.config!!.directory
    }

    fun getConfig(args: Args?): LogFileConfiguration? {
        return Database.log.file.config
    }

    fun setPlainTextStatus(args: Args): LogFileConfiguration {
        val config = args.get<LogFileConfiguration>("config")
        val plain_text = args.get<Boolean>("plain_text")
        config.setUsePlaintext(plain_text)
        return config
    }

    fun setMaxRotateCount(args: Args): LogFileConfiguration {
        val config = args.get<LogFileConfiguration>("config")
        val max_rotate_count = args.get<Int>("max_rotate_count")
        config.maxRotateCount = max_rotate_count
        return config
    }

    fun setMaxSize(args: Args): LogFileConfiguration {
        val config = args.get<LogFileConfiguration>("config")
        val max_size = args.get<Long>("max_size")
        config.maxSize = max_size
        return config
    }

    fun setConfig(args: Args): LogFileConfiguration {
        var directory = args.get<String>("directory")
        if (directory.isEmpty()) {
            val ts = System.currentTimeMillis() / 1000
            directory =
                RequestHandlerDispatcher.context.filesDir.absolutePath + File.separator + "logs_" + ts
            Log.i(
                TAG,
                "File logging configured at: $directory"
            )
        }
        val config = LogFileConfiguration(directory)
        Database.log.file.config = config
        return config
    }

    fun getLogLevel(args: Args?): Int {
        return Database.log.file.level.ordinal
    }

    fun setLogLevel(args: Args): LogFileConfiguration {
        val config = args.get<LogFileConfiguration>("config")
        val log_level = args.get<String>("log_level")
        when (log_level) {
            "debug" -> Database.log.file.level = LogLevel.DEBUG
            "verbose" -> Database.log.file.level = LogLevel.VERBOSE
            "info" -> Database.log.file.level = LogLevel.INFO
            "error" -> Database.log.file.level = LogLevel.ERROR
            "warning" -> Database.log.file.level = LogLevel.WARNING
            else -> Database.log.file.level = LogLevel.NONE
        }
        return config
    }

    fun getLogsInZip(args: Args?): RawData? {
        val fileLoggerConfig = Database.log.file.config ?: return null
        val zipper = ZipUtils()
        val zipDir = RequestHandlerDispatcher.context.getExternalFilesDir("zip")
        return try {
            val zipFile = File(zipDir, "archive.zip")
            if (zipFile.exists()) {
                zipper.deleteRecursive(zipFile)
            }
            zipper.zipDirectory(fileLoggerConfig.directory, zipFile)
            RawData("application/zip", zipper.readFile(zipFile))
        } finally {
            zipper.deleteRecursive(zipDir)
        }
    }

    companion object {
        private const val TAG = "LOGREQHANDLER"
    }
}